### PR TITLE
Refactor Ingredient to consist of sets

### LINKED ISSUE #58: Ingredients should be able to exist as multiple "sets"
For example, I have 500ml of milk expiring tomorrow, and 1.5l of milk expiring next week; these two should be treated as distinct units, but part of the same ingredient (milk).

When "making" a recipe, the program should always prefer to use the closest-expiring item (for obvious reasons)

### PR DESCRIPTION
**Pull request by zhiayang**
Closes #58

### CODE REVIEW

**Comment by zhiayang** [comment]
ready to merge

File: src/main/java/chopchop/model/ingredient/Ingredient.java [hunk 1]
```diff
@@ -33,16 +48,51 @@ public Ingredient(String name, Quantity quantity) {
...
+        super(name);
+        this.sets = sets;
     }
 
     public Quantity getQuantity() {
-        return this.quantity;
+        assert !this.sets.isEmpty();
+
+        // we *COULD* make an "identity" for Quantity, but that's too much effort, and each
+        // class of Quantity would need to explicitly handle that. Sadge.
```
Comment (Trav1sT) [inline src/main/java/chopchop/model/ingredient/Ingredient.java:68]: Agreed. Roger this.

File: src/main/java/chopchop/storage/JsonAdaptedIngredient.java [hunk 2]
```diff
@@ -51,28 +45,10 @@ public Ingredient toModelType() throws IllegalValueException {
...
-        if (modelQuantity.isError()) {
-            throw new IllegalValueException(modelQuantity.getError());
-        }
-
-        if (this.expiryDate == null) {
-            throw new IllegalValueException(String.format(INGREDIENT_MISSING_FIELD_MESSAGE_FORMAT,
-                ExpiryDate.class.getSimpleName()));
+        if (this.sets == null) {
+            throw new IllegalValueException(String.format(INGREDIENT_MISSING_FIELD_MESSAGE_FORMAT, "sets"));
         }
```
Comment (Trav1sT) [inline src/main/java/chopchop/storage/JsonAdaptedIngredient.java:50]: Concise!

File: src/main/java/chopchop/model/ingredient/Ingredient.java [hunk 3]
```diff
@@ -54,15 +104,37 @@ public Quantity getQuantity() {
...
-        return this.quantity.add(other.quantity)
-            .map(newQty -> new Ingredient(this.name.toString(), newQty, this.expiryDate))
-            .orElseThrow(IncompatibleIngredientsException::new);
+        // there's no constructor that takes both an existing map and the comparator...
+        var newSets = new TreeMap<Optional<ExpiryDate>, Quantity>(SET_COMPARATOR);
+        newSets.putAll(this.sets);
+
+        // because of exceptions, we cannot do this using nice lambdas and stuff.
+        // so write some dirty imperative code to merge the ingredients.
+        for (var entry : other.sets.entrySet()) {
```
Comment (Trav1sT) [inline src/main/java/chopchop/model/ingredient/Ingredient.java:118]: I'm just confused. What's the advantage of using `var` in general? Like in this case, wouldn't specifying the type make it clearer?

File: src/test/java/chopchop/storage/JsonAdaptedIngredientTest.java [hunk 4]
```diff
@@ -21,56 +26,77 @@
...
     @Test
-    public void toModelType_invalidEmail_throwsIllegalValueException() {
-        JsonAdaptedIngredient ingredient =
-            new JsonAdaptedIngredient(VALID_NAME, VALID_QTY, INVALID_EXPIRY);
-        String expectedMessage = ExpiryDate.MESSAGE_CONSTRAINTS;
+    public void toModelType_invalidExpiry_throwsIllegalValueException() {
+        var ingredient = new JsonAdaptedIngredient(VALID_NAME,
+            new JsonAdaptedIngredientSet(Arrays.asList(
+                new JsonAdaptedIngredientSet.JsonAdaptedPair(VALID_QTY, INVALID_EXPIRY)))
+            );
```
Comment (Trav1sT) [inline src/test/java/chopchop/storage/JsonAdaptedIngredientTest.java:84]: Beautiful! p.s. just 'reviewing'

**Review by Trav1sT — COMMENTED** [review]
Looks good!

File: src/main/java/chopchop/model/ingredient/Ingredient.java (hunk 3 above)
Comment (zhiayang) [reply src/main/java/chopchop/model/ingredient/Ingredient.java:118]: less typing. as you know typing places strain on your wrist, which is bad for health <-> `var` is good for health

ok but most of the time, `var` is used either (a) when the type is obvious, like this:
```java
var foo = new SomeOverlyLongClassName();
// versus:
SomeOverlyLongClassName foo = new SomeOverlyLongClassName();
```

and (b) when the type is not important, like in this case. explicitly writing `Map.Entry<Optional<ExpiryDate>, Quantity>` is tedious and doesn't contribute anything to the clarity of the code.

Comment (Trav1sT) [reply src/main/java/chopchop/model/ingredient/Ingredient.java:118]: Oo Thanks. This is more helpful than my doctor who only told me not to use it if it hurts but never say how.

**Pull request merged**